tests/generic_config_updater/test_pg_headroom_update.py

```diff
@@ -0,0 +1,111 @@
+from tests.generic_config_updater.gu_utils import create_checkpoint, delete_checkpoint, rollback_or_reload
+from tests.generic_config_updater.gu_utils import is_valid_platform_and_version
+
+pytestmark = [
+    pytest.mark.topology('any'),
+]
+
+logger = logging.getLogger(__name__)
+
+READ_ASICDB_TIMEOUT = 20
+READ_ASICDB_INTERVAL = 5
+
+
+@pytest.fixture(scope="function")
+def ensure_dut_readiness(duthost):
+    """
+    Setup/teardown fixture for pg headroom update test
+
+    Args:
+        duthost: DUT host object
+    """
+    verify_orchagent_running_or_assert(duthost)
+    create_checkpoint(duthost)
+
+    yield
+
+    try:
+        verify_orchagent_running_or_assert(duthost)
+        logger.info("Rolled back to original checkpoint")
+        rollback_or_reload(duthost)
+    finally:
+        delete_checkpoint(duthost)
+
+
+def ensure_application_of_updated_config(duthost, xoff, values):
+    """
+    Ensures application of the JSON patch config update
+
+    Args:
+        duthost: DUT host object
+        values: expected value(s) for the xoff field
+    """
+    def _confirm_value_in_app_and_asic_db():
+
+        for profile in xoff:
+            profile_data = duthost.shell('sonic-db-cli APPL_DB hgetall "BUFFER_PROFILE_TABLE:{}"'
+                                         .format(profile))["stdout"]
+            profile_data = ast.literal_eval(profile_data)
+            if profile_data["xoff"] != xoff[profile]:
+                return False
+
+        count = 0
+        table_name = duthost.shell('sonic-db-cli ASIC_DB keys *BUFFER_PROFILE*')["stdout_lines"]
+        for table in table_name:
+            profile_data = duthost.shell('sonic-db-cli ASIC_DB hgetall "{}"'.format(table))["stdout"]
+            profile_data = ast.literal_eval(profile_data)
+            if "SAI_BUFFER_PROFILE_ATTR_XOFF_TH" in profile_data:
+                count += 1
+                if profile_data["SAI_BUFFER_PROFILE_ATTR_XOFF_TH"] not in values:
+                    return False
+        return count == len(values)
+
+    logger.info("Validating fields in APPL DB and ASIC DB...")
+    pytest_assert(
+        wait_until(READ_ASICDB_TIMEOUT, READ_ASICDB_INTERVAL, 0, _confirm_value_in_app_and_asic_db),
+        "APPL DB or ASIC DB does not properly reflect newly configured value(s) for xoff"
+    )
+
+
+@pytest.mark.parametrize("operation", ["replace"])
+def test_pg_headroom_update(duthost, ensure_dut_readiness, operation):
+    tmpfile = generate_tmpfile(duthost)
+
+    json_patch = list()
+    values = list()
+    xoff = dict()
+    lossless_profiles = duthost.shell('sonic-db-cli CONFIG_DB keys *BUFFER_PROFILE\\|pg_lossless*')['stdout_lines']
+    for profile in lossless_profiles:
+        profile_name = profile.split('|')[-1]
+        value = duthost.shell('sonic-db-cli CONFIG_DB hget "{}" "xoff"'.format(profile))['stdout']
+        value = ast.literal_eval(value)
+        value -= 1000
```

Comment:
Will it become a negative number after - 1000?

Reply:
No, the values are always in 10K range